File: packages/camera/camera_avfoundation/example/ios/RunnerTests/Mocks/MockCaptureDevice.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "MockCaptureDevice.h"

@import camera_avfoundation;
#if __has_include(<camera_avfoundation/camera_avfoundation-umbrella.h>)
@import camera_avfoundation.Test;
#endif
@import AVFoundation;

@implementation MockCaptureDevice

- (NSObject<FLTCaptureDeviceFormat> *)activeFormat {
  if (self.activeFormatStub) {
    return self.activeFormatStub();
  }
  return nil;
}

- (void)setActiveFormat:(NSObject<FLTCaptureDeviceFormat> *)format {
  if (self.setActiveFormatStub) {
    self.setActiveFormatStub(format);
  }
}

- (BOOL)isFlashModeSupported:(AVCaptureFlashMode)mode {
  return self.flashModeSupported;
}

- (void)setTorchMode:(AVCaptureTorchMode)mode {
  if (self.setTorchModeStub) {
    self.setTorchModeStub(mode);
  }
}

- (BOOL)isFocusModeSupported:(AVCaptureFocusMode)mode {
  if (self.isFocusModeSupportedStub) {
    return self.isFocusModeSupportedStub(mode);
  }
  return NO;
}

- (void)setFocusMode:(AVCaptureFocusMode)mode {
  if (self.setFocusModeStub) {
    self.setFocusModeStub(mode);
  }
}

- (void)setFocusPointOfInterest:(CGPoint)point {
  if (self.setFocusPointOfInterestStub) {
    self.setFocusPointOfInterestStub(point);
  }
}

- (void)setExposureMode:(AVCaptureExposureMode)mode {
  if (self.setExposureModeStub) {
    self.setExposureModeStub(mode);
  }
}

- (void)setExposurePointOfInterest:(CGPoint)point {
  if (self.setExposurePointOfInterestStub) {
    self.setExposurePointOfInterestStub(point);
  }
}

- (void)setExposureTargetBias:(float)bias completionHandler:(void (^)(CMTime))handler {
  if (self.setExposureTargetBiasStub) {
    self.setExposureTargetBiasStub(bias, handler);
  }
}

- (void)setVideoZoomFactor:(float)factor {
  if (self.setVideoZoomFactorStub) {
    self.setVideoZoomFactorStub(factor);
  }
}

- (BOOL)lockForConfiguration:(NSError **)error {
  if (self.lockForConfigurationStub) {
    return self.lockForConfigurationStub(error);
  }
  return YES;
}

- (void)unlockForConfiguration {
  if (self.unlockForConfigurationStub) {
    self.unlockForConfigurationStub();
  }
}

- (void)setActiveVideoMinFrameDuration:(CMTime)duration {
  if (self.setActiveVideoMinFrameDurationStub) {
    self.setActiveVideoMinFrameDurationStub(duration);
  }
}

- (void)setActiveVideoMaxFrameDuration:(CMTime)duration {
  if (self.setActiveVideoMaxFrameDurationStub) {
    self.setActiveVideoMaxFrameDurationStub(duration);
  }
}

- (BOOL)isExposureModeSupported:(AVCaptureExposureMode)mode {
  return self.exposureModeSupported;
}

@synthesize device;

@end

@implementation MockCaptureInput
@synthesize ports;
@synthesize input;
@end

@implementation MockCaptureDeviceInputFactory

- (nonnull instancetype)init {
  self = [super init];
  if (self) {
    _mockDeviceInput = [[MockCaptureInput alloc] init];
  }
  return self;
}

- (nonnull instancetype)initWithMockDeviceInput:
    (nonnull NSObject<FLTCaptureInput> *)mockDeviceInput {
  self = [super init];
  if (self) {
    _mockDeviceInput = mockDeviceInput;
  }
  return self;
}

- (NSObject<FLTCaptureInput> *)deviceInputWithDevice:(NSObject<FLTCaptureDevice> *)device
                                               error:(NSError **)error {
  return _mockDeviceInput;
}

@end

File: packages/camera/camera_avfoundation/example/ios/RunnerTests/Mocks/MockCaptureDevice.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

@testable import camera_avfoundation

// Import Objective-C part of the implementation when SwiftPM is used.
#if canImport(camera_avfoundation_objc)
  import camera_avfoundation_objc
#endif

/// A mock implementation of `FLTCaptureDevice` that allows mocking the class
/// properties.
class MockCaptureDevice: NSObject, CaptureDevice {
  var activeFormatStub: (() -> CaptureDeviceFormat)?
  var setActiveFormatStub: ((CaptureDeviceFormat) -> Void)?
  var getTorchModeStub: (() -> AVCaptureDevice.TorchMode)?
  var setTorchModeStub: ((AVCaptureDevice.TorchMode) -> Void)?
  var isFocusModeSupportedStub: ((AVCaptureDevice.FocusMode) -> Bool)?
  var setFocusModeStub: ((AVCaptureDevice.FocusMode) -> Void)?
  var setFocusPointOfInterestStub: ((CGPoint) -> Void)?
  var setExposureModeStub: ((AVCaptureDevice.ExposureMode) -> Void)?
  var setExposurePointOfInterestStub: ((CGPoint) -> Void)?
  var setExposureTargetBiasStub: ((Float, ((CMTime) -> Void)?) -> Void)?
  var isExposureModeSupportedStub: ((AVCaptureDevice.ExposureMode) -> Bool)?
  var setVideoZoomFactorStub: ((CGFloat) -> Void)?
  var lockForConfigurationStub: (() throws -> Void)?

  var avDevice: AVCaptureDevice {
    preconditionFailure("Attempted to access unimplemented property: device")
  }

  var uniqueID = ""
  var position = AVCaptureDevice.Position.unspecified
  var deviceType = AVCaptureDevice.DeviceType.builtInWideAngleCamera

  var flutterActiveFormat: CaptureDeviceFormat {
    get {
      activeFormatStub?() ?? MockCaptureDeviceFormat()
    }
    set {
      setActiveFormatStub?(newValue)
    }
  }

  var flutterFormats: [CaptureDeviceFormat] = []
  var hasFlash = false
  var hasTorch = false
  var isTorchAvailable = false
  var torchMode: AVCaptureDevice.TorchMode {
    get {
      getTorchModeStub?() ?? .off
    }
    set {
      setTorchModeStub?(newValue)
    }
  }
  var isFocusPointOfInterestSupported = false
  var maxAvailableVideoZoomFactor = CGFloat(0)
  var minAvailableVideoZoomFactor = CGFloat(0)
  var videoZoomFactor: CGFloat {
    get { 0 }
    set {
      setVideoZoomFactorStub?(newValue)
    }
  }
  var isExposurePointOfInterestSupported = false
  var minExposureTargetBias = Float(0)
  var maxExposureTargetBias = Float(0)
  var activeVideoMinFrameDuration = CMTime(value: 1, timescale: 1)
  var activeVideoMaxFrameDuration = CMTime(value: 1, timescale: 1)

  func isFlashModeSupported(_ mode: AVCaptureDevice.FlashMode) -> Bool {
    return false
  }

  func isFocusModeSupported(_ mode: AVCaptureDevice.FocusMode) -> Bool {
    return isFocusModeSupportedStub?(mode) ?? false
  }

  var focusMode: AVCaptureDevice.FocusMode {
    get { .autoFocus }
    set { setFocusModeStub?(newValue) }
  }

  func setFocusMode(_ focusMode: AVCaptureDevice.FocusMode) {
    setFocusModeStub?(focusMode)
  }

  var focusPointOfInterest: CGPoint {
    get { CGPoint.zero }
    set { setFocusPointOfInterestStub?(newValue) }
  }

  var exposureMode: AVCaptureDevice.ExposureMode {
    get { .autoExpose }
    set { setExposureModeStub?(newValue) }
  }

  var exposurePointOfInterest: CGPoint {
    get { CGPoint.zero }
    set { setExposurePointOfInterestStub?(newValue) }
  }

  func setExposureTargetBias(_ bias: Float, completionHandler handler: ((CMTime) -> Void)? = nil) {
    setExposureTargetBiasStub?(bias, handler)
  }

  func isExposureModeSupported(_ mode: AVCaptureDevice.ExposureMode) -> Bool {
    return isExposureModeSupportedStub?(mode) ?? false
  }

  var lensAperture: Float { 0 }

  var exposureDuration: CMTime { CMTime(value: 1, timescale: 1) }

  var iso: Float { 0 }

  func lockForConfiguration() throws {
    try lockForConfigurationStub?()
  }

  func unlockForConfiguration() {}
}
